File: MatlabCode/Chapter3_Time-independentSE/Exercise_3_4_2.m
% This script estimates the ground state energy of a "smooth" 
% rectangular potential by means of the variational principle. 
% As a trial function we use a Gaussian wave packet centered at the 
% origin with a variable width.
%
% The Hamiltonian is estimated numerically using an FFT representation 
% of the kinetic energy.
%
% The ground state energy is also calculated "exactly" by diagoalizing 
% the full, numerical Hamiltonian.
%
% The estimated energy is plotted as a function of the width - along 
% with the "exact" ground state energy
%
% Numerical inputs:
% SigmaMin  - minimal value for the widht
% SigmaMax  - maximal value for the width
% SigmaStep - step size used for the width
% L         - size of domain 
% N         - number of grid points (should be 2^n)
%
% Physical inputs:
% V0        - the height of the potential (should be negative)
% w         - the width of the potential
% s         - the "smoothness" of the potential 
% Vpot      - the actual potential (function variable)
%
% All parameters and functions are hard coded initially

SigmaMin = 0.1;
SigmaMax = 5;
SigmaStep = 0.05;

% Trial function
PsiTrial = @(x,sigma) (2*pi*sigma^2)^(-0.25)*exp(-x.^2/(4*sigma^2));

% Potential:
V0 = -3;
w = 8;
s = 5;

% Grid parameters
L = 30;
N = 1024;              % For FFT's sake, we should have N=2^n

% Shape of the potential
Vpot = @(x) V0./(exp(s*(abs(x)-w/2))+1);

% Set up the grid.
x = linspace(-L/2,L/2,N);
x = x.';                            % Traspose
h = L/(N-1);

% Set up kinetic energy matrix by means of the fast Fourier transform
k = 2*pi/(N*h)*[0:(N/2-1), (-N/2:-1)];          % Vector with k-values
% Fourier transform the identity matrix:
Tmat_FFT = fft(eye(N));
% Multiply by (ik)^2
Tmat_FFT = diag(-k.^2)*Tmat_FFT;
% Transform back to x-representation
Tmat_FFT = ifft(Tmat_FFT);
Tmat_FFT = -1/2*Tmat_FFT;            % Correct prefactor

% Total Hamiltonian
H = Tmat_FFT + diag(Vpot(x));

% Calculate expectation value of the Hamiltonian
% for various sigmas
index = 1;
for sigma = SigmaMin:SigmaStep:SigmaMax
  Psi = PsiTrial(x,sigma);                 % Wave function
  Sigma(index) = sigma;                   % Vector with sigma-values
  Energy(index) = h*Psi'*H*Psi;           % Expectation value
  index=index+1;                          % Update index
end

% Diagonalize to find eigen states and energies
[EigVectors EigValues] = eig(H);
% Extract eigen values from diagonal matrix
EigValues = diag(EigValues);
% Sort them (and remove spurious imaginary parts)
[EigValues, Indexes] = sort(real(EigValues));
EigVectors = EigVectors(:,Indexes);
% "Exact" groundState
GroundStateEnergy = EigValues(1);
GroundStateWF = EigVectors(:,1)/sqrt(h);  % Normalize

% Determine estimate - and error
[EnergyEstimate, MinInd] = min(Energy);
Error = EnergyEstimate - GroundStateEnergy;
disp(['Ground state energy estimate: ', ...
    num2str(EnergyEstimate)])
disp(['Actual ground state energy (on the grid): ', ...
    num2str(GroundStateEnergy)])
disp(['The error in ground state energy is ', num2str(Error),'.'])
disp(['This is corresponds to the relative error ',...
    num2str(abs(Error/GroundStateEnergy)*100),' %.'])

%
% Plot ground state energy estimate and compare with "exact" result
%
figure(1)
plot(Sigma, Energy, 'k-', 'linewidth', 2)
hold on
yline(GroundStateEnergy, 'r--', 'linewidth', 2);
hold off
grid on
set(gca,'fontsize',15)
xlabel('\sigma')
legend('<E(\sigma)>','Exact')

%
% Plot wave functions - the estimate and the "exact" one
%
% Estimated ground state
Psi = PsiTrial(x, Sigma(MinInd));               
figure(2)
plot(x, abs(Psi).^2, 'k-', 'linewidth', 2)
hold on
% Plott "exact" wave function
plot(x, abs(GroundStateWF).^2, 'r--', 'linewidth', 2)          
% Fix strength of the potential for plotting
MaxWalPot = -.2*max(abs(Psi)).^2/V0;          
% Plot confining potential
plot(x, Vpot(x)*MaxWalPot, 'b');      
hold off
grid on
set(gca,'fontsize',15)
legend('Estimate','Exact','Potential')

File: PythonCode/Chapter3_Time-indepdentSE/Exercise_3_4_2.py
"""
 This script estimates the ground state energy of a "smooth" rectangular 
 potential by means of the variational principle. As a trial function we 
 use a Gaussian wave packet centered at the origin with a variable width.

 The Hamiltonian is estimated numerically using an FFT representation of 
 the kinetic energy.

 The ground state energy is also calculated "exactly" by diagoalizing the 
 full, numerical Hamiltonian.

 The estimated energy is plotted as a function of the width - along with 
 the "exact" ground state energy

 Numerical inputs:
 SigmaMin  - minimal value for the widht
 SigmaMax  - maximal value for the width
 SigmaStep - step size used for the width
 L         - size of domain 
 N         - number of grid points (should be 2^n)

 Physical inputs:
 V0        - the height of the potential (should be negative)
 w         - the width of the potential
 s         - the "smoothness" of the potential 
 Vpot      - the actual potential (function variable)

 All parameters and functions are hard coded initially
"""

# Libraries
import numpy as np
from matplotlib import pyplot as plt

# Grid for the sigma parameter
SigmaMin = 0.1
SigmaMax = 5
SigmaStep = 0.05

# Numerical grid parameters
L = 30
N = 2048                


# Inputs for the smoothly rectangular potential
V0 = -3            # Height, i.e. negative debth
w = 8              # Width
s = 5              # Smootness

# Potential (as function)
def Vpot(x):
    return V0/(np.exp(s*(np.abs(x)-w/2))+1.0)

# Set up grids
# x-grid
x = np.linspace(-L/2, L/2, N)
h = L/(N-1)
# sigma-grid
Nsigma = int(np.floor((SigmaMax-SigmaMin)/SigmaStep))+1
SigmaVector = np.linspace(SigmaMin, SigmaMax, Nsigma)

# Trial function
def PsiFunk(x, Sigma):
    return (2*np.pi*Sigma**2)**(-0.25)*np.exp(-x**2/(4*Sigma**2))

# Kinetic energy operator (FFT)
k_max = np.pi/h
dk = 2*k_max/N
k = np.append(np.linspace(0, k_max-dk, int(N/2)), 
              np.linspace(-k_max, -dk, int(N/2)))
# Transform identity matrix
Tmat_FFT = np.fft.fft(np.identity(N, dtype=complex), axis = 0)
# Multiply by (ik)^2
Tmat_FFT = np.matmul(np.diag(-k**2), Tmat_FFT)
# Transform back to x-representation. 
Tmat_FFT = np.fft.ifft(Tmat_FFT, axis = 0)
# Correct pre-factor (unit mass and \hbar = 1)
Tmat_FFT = -1/2*Tmat_FFT    

# Full Hamiltonian
Ham = Tmat_FFT + np.diag(Vpot(x))
# Ensure real Hamiltonian
Ham = np.real(Ham)

# Initiate and allocate
Sigma = SigmaMin
SigmaInd = 0
EnergyMin = 1e6
EnergyVector = np.zeros(Nsigma)
for Sigma in SigmaVector:
    # Trial function
    Psi = PsiFunk(x, Sigma)
    # Hermitian adjoint
    PsiDagger = np.transpose(Psi)
    PsiDagger = np.conj(PsiDagger)
    # Energy expectation value
    Energy = h*np.matmul(PsiDagger, np.matmul(Ham, Psi))
    # Check for new minimum
    if Energy < EnergyMin:
        EnergyMin = Energy
        SigmaForMinimum = Sigma
    # Write energy to vector
    EnergyVector[SigmaInd] = Energy
    # Update
    Sigma = Sigma + SigmaStep
    SigmaInd = SigmaInd + 1        
    
# Diagaonalize Hamiltonian (Hermitian matrix)
Evector, PsiMat = np.linalg.eigh(Ham)
# Extract ground state
GroundStateEnergy = Evector[0]
GroundState = PsiMat[:,0]
# Normalize
GroundState = PsiMat[:,0]
GroundState = GroundState/np.sqrt(h)

# Write restulst to screen
print(f'Minimal energy: {EnergyMin:.4f}')
print(f'Actual ground state energy: {GroundStateEnergy:.4f}')
Error = EnergyMin-GroundStateEnergy
print(f'Error: {Error:.4f}')
RelError = Error/np.abs(GroundStateEnergy)
print(f'Relative error: {RelError*100:2.2f} %')


# Plot energy expectation value
plt.figure(1)
plt.clf()
plt.plot(SigmaVector, EnergyVector, 'k-', label = '<E($\sigma$)>')
plt.axhline(y = GroundStateEnergy, color = 'r', 
            linestyle = '--', label = 'Actual energy')
plt.xlabel('$\sigma$')
plt.legend()
plt.grid()
plt.show()

# Plot wave functions 
# - both the one that minimizes E and the real one
plt.figure(2)
plt.clf()
# Variational
PsiVar = PsiFunk(x, SigmaForMinimum)
plt.plot(x, np.abs(PsiVar)**2, 'k-', label = 'Variatonal')
# Actual
plt.plot(x, np.abs(GroundState)**2, 'r--', label = 'Actual')
# Potential
ScaleFactor = np.max(np.abs(PsiVar)**2)
ScaleFactor = 0.2*ScaleFactor/np.abs(V0)
plt.plot(x, ScaleFactor*Vpot(x), 'b', label = 'Potential')
# Cosmetics
plt.xlabel('x', fontsize = 12)
plt.legend()
plt.grid()
plt.show()
